flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/rules/logical/PushPartitionIntoTableSourceScanRule.java

```diff
@@ -0,0 +1,320 @@
+					"Validator should have checked that.", name));
+			}
+			return inputFieldTypes.getFieldList().get(index).getType(); })
+			.map(FlinkTypeFactory::toLogicalType).collect(Collectors.toList());
+
+		// get partitions from table source and prune
+		List<Map<String, String>> remainingPartitions = null;
+		Optional<List<Map<String, String>>> optionalPartitions = null;
+		try {
+			optionalPartitions = ((SupportsPartitionPushDown) dynamicTableSource).listPartitions();
+		} catch (UnsupportedOperationException e) {
+			// read partitions from catalog if table source doesn't support listPartitions operation.
+			if (!catalogOptional.isPresent()){
+				throw new TableException(
+					String.format("Table %s must from a catalog, but %s is not a catalog",
+						identifier.asSummaryString(), identifier.getCatalogName()), e);
+			}
+		}
+		if (optionalPartitions != null) {
+			if (!optionalPartitions.isPresent() || optionalPartitions.get().size() == 0) {
+				// return if no partitions
+				return;
+			}
+			// get remaining partitions
+			remainingPartitions = internalPrunePartitions(
+				optionalPartitions.get(),
+				inputFieldNames,
+				partitionFieldNames,
+				partitionFieldTypes,
+				partitionPredicate,
+				context.getTableConfig());
+		} else {
+			RexNodeToExpressionConverter converter = new RexNodeToExpressionConverter(
+				inputFieldNames.toArray(new String[0]),
+				context.getFunctionCatalog(),
+				context.getCatalogManager(),
+				TimeZone.getTimeZone(context.getTableConfig().getLocalTimeZone()));
+			ArrayList<Expression> exprs = new ArrayList<>();
+			Option<ResolvedExpression> subExpr;
+			for (RexNode node: JavaConversions.seqAsJavaList(allPredicates._1)) {
+				subExpr = node.accept(converter);
+				if (!subExpr.isEmpty()) {
+					exprs.add(subExpr.get());
+				}
+			}
+			try {
+				if (exprs.size() > 0) {
+					remainingPartitions = catalogOptional.get().listPartitionsByFilter(tablePath, exprs)
+						.stream()
+						.map(CatalogPartitionSpec::getPartitionSpec)
+						.collect(Collectors.toList());
+				} else {
+					// no filter and get all partitions
+					List<Map<String, String>> partitions = catalogOptional.get().listPartitions(tablePath)
+						.stream()
+						.map(CatalogPartitionSpec::getPartitionSpec)
+						.collect(Collectors.toList());
+					// prune partitions
+					if (partitions.size() > 0) {
+						remainingPartitions = internalPrunePartitions(
+							partitions,
+							inputFieldNames,
+							partitionFieldNames,
+							partitionFieldTypes,
+							partitionPredicate,
+							context.getTableConfig());
+					} else {
+						return;
+					}
+				}
+			} catch (TableNotExistException e) {
+				throw new TableException(String.format("Table %s is not found in catalog.", identifier.asSummaryString()), e);
+			} catch (TableNotPartitionedException e) {
+				// no partitions in table source
+				return;
+			}
+		}
```

Comment:
I think it's better we can split the logic to different sub-methods, and simplify this part logic as:
1. get partition from TableSource, if succeed, do partition pruning and build new table scan, else fallback to step 2
2. check whether the catalog exists. if not existing, return. else go to step 2.1
2.1. try to get partitions through `listPartitionsByFilter` method. if succeed, build new table scan. else go to step 2.2
2.2. try to get partitions through `listPartitions` method. if failed, return, else do partition pruning and build new table scan.